pkg/scheduler/core/equivalence/eqivalence.go

```diff
@@ -33,22 +33,146 @@ import (
 	"github.com/golang/glog"
 )
 
-// Cache saves and reuses the output of predicate functions. Use RunPredicate to
+// nodeMap stores a *Cache for each node.
+type nodeMap map[string]*NodeCache
+
+// Cache is a thread safe nodeMap
+// NOTE(harry): Theoretically sync.Map has better performance in machine with 8+ CPUs, while
+// the reality is lock contention in first level cache is rare.
+type Cache struct {
+	mu          sync.RWMutex
+	nodeToCache nodeMap
+}
+
+// NewCache create am empty equiv class cache.
+func NewCache() *Cache {
+	return &Cache{
+		nodeToCache: make(nodeMap),
+	}
+}
+
+// ForNode returns the existing NodeCache for given node if present. Otherwise,
+// it creates the NodeCache and returns it.
+// The boolean flag is true if the value was loaded, false if created.
+func (n *Cache) ForNode(name string) (nodeCache *NodeCache, exists bool) {
```

Comment:
What should callers do differently depending on the boolean return?

Reply:
I think, it would be nice to expose some indication of the change we did in `ForNode()` call as it has mix duties. This is only used by tests for now.